src/envoy/http/service_control/handler_utils.cc

```diff
@@ -359,6 +364,55 @@ void fillStatus(const Envoy::Http::ResponseHeaderMap* response_headers,
   info.grpc_response_code = static_cast<StatusCode>(status.value());
 }
 
+std::string extractIPFromForwardedHeader(
+    const Envoy::Http::RequestHeaderMap& headers) {
+  const auto values = headers.get(kForwardedHeader);
+
+  // Only support one header value.
+  // Need to extract the last one,  could not define the last one for multiple
+  // header values.
+  if (values.size() != 1) {
+    return Envoy::EMPTY_STRING;
+  }
+
+  absl::string_view source = values[0]->value().getStringView();
+
+  // Multiple proxy sections are separated by comma. Use the last one.
+  const absl::string_view::size_type pos = source.find_last_of(',');
+  if (pos != absl::string_view::npos) {
+    source.remove_prefix(pos + 1);
+  }
+
+  // Each section may have multiple fields, they are separated by semicolumn.
+  // For example: by=abc;for=1.2.3.4;host=abc;proto=http
+  // Extract ip from the "for=" field.
+  for (absl::string_view s : absl::StrSplit(source, ';')) {
```

Comment:
probably should extract out all magic strings and numbers into constants

Reply:
Updated for "for=",  but not for single character which is more readable here.